arelle/plugin/validate/ROS/rules/ros.py

```diff
@@ -338,3 +337,40 @@ def rule_ros18(
             code='ROS.18',
             message=_("'%(conceptLn)s' cannot have a negative value.")
         )
+
+
+@validation(
+    hook=ValidationHook.XBRL_FINALLY,
+)
+def rule_ros19(
+        pluginData: PluginValidationDataExtension,
+        val: ValidateXbrl,
+        *args: Any,
+        **kwargs: Any,
+) -> None:
+    """
+    ROS: Rule 19:PrincipalCurrencyUsedInBusinessReport must exist and its value must match the name of the unit used for all monetary facts.
+    """
+    message: str | None = None
+    modelObjects = set()
+    monetaryFacts = val.modelXbrl.factsByDatatype(False, qnXbrliMonetaryItemType)
+    monetaryUnits = set(fact.unit.value for fact in monetaryFacts)
+    pricipalCurrencyFacts = val.modelXbrl.factsByLocalName.get(PRINCIPAL_CURRENCY, set())
+    principalCurrencyValues = set(fact.text for fact in pricipalCurrencyFacts)
+    if len(principalCurrencyValues) != 1:
+        modelObjects = pricipalCurrencyFacts
+        message = _("'%(conceptName)s' must exist and have a single value.  Values found: %(principalCurrencyValues)s.")
+    elif monetaryUnits != principalCurrencyValues:
+        for fact in monetaryFacts:
+            if fact.unit.value != principalCurrencyValues:
```

Comment:
Isn't `fact.unit.value` a string while `principalCurrencyValues` is a list of strings? So I think all `monetaryFacts` will be added to `modelObjects` since this expression will always be `True`.